File: backend/services/resume_storage_policy.py
from __future__ import annotations

from typing import Any, Dict


_ANALYSIS_SOURCES = {
    'diagnosis_generated',
    'analysis_generated',
    'interview_refined',
}


def _to_text(value: Any) -> str:
    return str(value or '').strip()


def _to_normalized(value: Any) -> str:
    return _to_text(value).lower()


def _is_non_empty_dict(value: Any) -> bool:
    return isinstance(value, dict) and len(value) > 0


def is_resume_eligible_for_library(resume_data: Dict[str, Any] | Any) -> bool:
    if not isinstance(resume_data, dict):
        return False

    optimization_status = _to_normalized(resume_data.get('optimizationStatus'))
    if optimization_status == 'optimized':
        return True

    if _to_text(resume_data.get('analysisReportId')):
        return True
    if _to_text(resume_data.get('optimizationJdKey')):
        return True

    if _is_non_empty_dict(resume_data.get('analysisSnapshot')):
        return True
    if _is_non_empty_dict(resume_data.get('analysisDossierLatest')):
        return True
    if _is_non_empty_dict(resume_data.get('analysisBindings')):
        return True
    if _is_non_empty_dict(resume_data.get('analysisSessionByJd')):
        return True

    source = _to_normalized(
        resume_data.get('source')
        or resume_data.get('resumeSource')
        or resume_data.get('generatedSource')
    )
    if source in _ANALYSIS_SOURCES:
        return True

    return False

File: ai-resume-builder/src/resume-storage-policy.ts
const ANALYSIS_SOURCES = new Set([
  'diagnosis_generated',
  'analysis_generated',
  'interview_refined',
]);

const toText = (value: unknown): string => String(value || '').trim();

const toNormalized = (value: unknown): string => toText(value).toLowerCase();

const isObjectRecord = (value: unknown): value is Record<string, any> =>
  !!value && typeof value === 'object' && !Array.isArray(value);

const hasNonEmptyKeys = (value: unknown): boolean =>
  isObjectRecord(value) && Object.keys(value).length > 0;

const toResumeData = (resumeLike: any): Record<string, any> => {
  if (isObjectRecord(resumeLike?.resume_data)) return resumeLike.resume_data;
  if (isObjectRecord(resumeLike)) return resumeLike;
  return {};
};

export const isResumeEligibleForLibrary = (resumeLike: any): boolean => {
  const resumeData = toResumeData(resumeLike);

  const optimizationStatus = toNormalized(resumeData.optimizationStatus);
  if (optimizationStatus === 'optimized') return true;

  if (toText(resumeData.analysisReportId)) return true;
  if (toText(resumeData.optimizationJdKey)) return true;

  if (hasNonEmptyKeys(resumeData.analysisSnapshot)) return true;
  if (hasNonEmptyKeys(resumeData.analysisDossierLatest)) return true;
  if (hasNonEmptyKeys(resumeData.analysisBindings)) return true;
  if (hasNonEmptyKeys(resumeData.analysisSessionByJd)) return true;

  const source = toNormalized(resumeData.source || resumeData.resumeSource || resumeData.generatedSource);
  if (ANALYSIS_SOURCES.has(source)) return true;

  return false;
};

export const shouldPersistResumeRecord = (resumeLike: any): boolean =>
  isResumeEligibleForLibrary(resumeLike);

export const withLocalOnlyDraftMeta = (resumeLike: any): Record<string, any> => {
  const resumeData = toResumeData(resumeLike);
  const source = toNormalized(resumeData.source) || 'profile_draft';
  return {
    ...resumeData,
    source,
    persistPolicy: 'local_only',
  };
};

